sunshine/video.cpp

```diff
@@ -474,6 +475,34 @@ static encoder_t amdvce {
   DEFAULT,
   dxgi_make_hwdevice_ctx
 };
+
+
+static encoder_t quicksync {
+  "quicksync"sv,
+  { FF_PROFILE_H264_HIGH, FF_PROFILE_HEVC_MAIN },
+  AV_HWDEVICE_TYPE_QSV,
+  AV_PIX_FMT_QSV,
+  AV_PIX_FMT_NV12, AV_PIX_FMT_P010,
+  {
+    {
+      { "forced-idr"s, 1 },
```

Comment:
According to https://stackoverflow.com/questions/66921104/how-to-force-qsv-generate-idr-when-using-qsv-encoding-in-ffmpeg  and https://gist.github.com/nico-lab/597edc466ec54d0b6ce9dbf263da766a , the option is actually `forced_idr`

Reply:
Yes, but have tried forced_idr as well. Actually I could manage it to send only idr frames but then getting:
00:10:10 - SDL Info (0): Unrecoverable frame 740637842: lost FEC blocks 1 to 3
00:10:10 - SDL Info (0): Unrecoverable frame -1801679895: lost FEC blocks 1 to 3
00:10:10 - SDL Info (0): Unrecoverable frame 1140981760: 0+2=2 received < 659 needed

I have tried many options to solve this "Waiting for IDR frame" error but none of them seemed to work

Reply:
If you capture the raw encoder output to a file, I can take a look at the bitstream and figure out what's going on.

Does it affect both H.264 and HEVC?

Reply:
@cgutman yep, both of them. I can send you the file

Reply:
not sure it would help you but I was comparing software encode bitstream against qsv and it looks like nal_ref_idc is different. For instance SPS, PPS and IDR are 1 on qsv while it is 3 on software bitstream.

Reply:
ok I finally got video and frames are running fine on moonlight
I basically needed to do this:

```
    payload_new = replace(payload, "\000\000\000\001\'"sv, "\000\000\000\001g"sv);
    payload     = { (char *)payload_new.data(), payload_new.size() };

    payload_new = replace(payload, "\000\000\000\001("sv, "\000\000\000\001h"sv);
    payload     = { (char *)payload_new.data(), payload_new.size() };

    payload_new = replace(payload, "\000\000\001%"sv, "\000\000\001e"sv);
    payload     = { (char *)payload_new.data(), payload_new.size() };
```

It replaces the nal_ref_idr of key frames to 3

Reply:
Very interesting. Old versions of Moonlight did care about `nal_ref_idc` (in violation of the H.264 and HEVC specification), but recent versions (within the last year or so) shouldn't care. The fix was https://github.com/moonlight-stream/moonlight-common-c/commit/b528867466fe780aa7b3f5309f4fd8dfa3b06765

I wonder if what's going on is that Sunshine itself is breaking the bitstream (or rather the fixups it's trying to perform aren't working because the original payload doesn't have the expected bytes).

Reply:
I looked at output.h264 and it looks like the issue is that the NALU for the IDR image has the 3-byte Annex B start code rather than the 4-byte Annex B start code that Moonlight expects for the start of a frame (and which is used for the SPS, PPS, and SEI NALUs).

![image](https://user-images.githubusercontent.com/2695644/156283132-7cab11fa-a7b0-4f21-adcd-b56c28632945.png)

I think using a 3-byte start code is legal, but just changing moonlight-common-c to accept would expose bugs in some clients (iOS at least) that assume the NALU start prefixes are always 4 bytes long. Handling both 3 and 4 byte start prefixes would complicate the logic we use on iOS to convert from Annex B to AVCC format. I think it makes sense to continue to fix this up on the Sunshine side, but we can make the logic more robust to avoid getting confused if the first NALU byte varies between decoders.

Reply:
yeah, makes total sense to me. thanks for you support.